modules/common/tls/tls.go

```diff
@@ -0,0 +1,192 @@
+}
+
+// NewTLS - initialize and return a TLS struct
+func NewTLS(cr client.Object, clientset *kubernetes.Clientset, scheme *runtime.Scheme, namespace string, service Service, ca Ca) (*TLS, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	crClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	if err != nil {
+		return nil, err
+	}
+
+	helperInstance, err := helper.NewHelper(cr, crClient, clientset, scheme, log.Log)
+	if err != nil {
+		return nil, fmt.Errorf("could not create helper instance: %w", err)
+	}
+
+	// Ensure service SecretName exists or return an error
+	if service.SecretName == "" {
+		service.SecretName = "default-tls-secret"
+	} else {
+		_, _, err := secret.GetSecret(context.TODO(), helperInstance, service.SecretName, namespace)
+		if err != nil {
+			return nil, fmt.Errorf("error ensuring secret %s exists: %w", service.SecretName, err)
+		}
+	}
+
+	// Ensure ca CaSecretName exists or return an error
+	if ca.CaSecretName == "" {
+		ca.CaSecretName = "default-ca-secret"
+	} else {
+		_, _, err := secret.GetSecret(context.TODO(), helperInstance, ca.CaSecretName, namespace)
+		if err != nil {
+			return nil, fmt.Errorf("error ensuring secret %s exists: %w", ca.CaSecretName, err)
+		}
+	}
+
+	return &TLS{
+		Service: service,
+		Ca:      ca,
+	}, nil
+}
+
+// CreateVolumeMounts - add volume mount for TLS certificate and CA certificates, this counts on openstack-operator providing CA certs with unique names
+func CreateVolumeMounts(tls *Service, ca *Ca) []corev1.VolumeMount {
+	var volumeMounts []corev1.VolumeMount
+
+	if tls.SecretName != "" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "tls-certs",
+			MountPath: "/var/tls-certs",
+			ReadOnly:  true,
+		})
+	}
+
+	if ca.CaSecretName != "" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "ca-certs",
+			MountPath: "/var/ca-certs",
+			ReadOnly:  true,
+		})
+	}
+
+	return volumeMounts
+}
+
+// CreateVolumes - add volume for TLS certificate and CA certificates
+func CreateVolumes(tls *Service, ca *Ca) []corev1.Volume {
+	var volumes []corev1.Volume
+	mode := int32(0400)
+
+	if tls.SecretName != "" {
+		volumes = append(volumes, corev1.Volume{
+			Name: "tls-certs",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  tls.SecretName,
+					DefaultMode: &mode,
+				},
+			},
+		})
+	}
+
+	if ca.CaSecretName != "" {
+		volumes = append(volumes, corev1.Volume{
+			Name: "ca-certs",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  ca.CaSecretName,
+					DefaultMode: &mode,
+				},
+			},
+		})
+	}
+
+	return volumes
+}
+
+// SetProbes sets the liveness and readiness probes based on the TLS configuration, disable HTTPS for debugging purposes
+func SetProbes(port int, disableNonTLSListeners bool) (livenessProbe *corev1.Probe, readinessProbe *corev1.Probe, err error) {
```

Comment:
this Probes func might be better located in a probes package instead of the tls here.

Reply:
we'd also need to pass in the the path v3 is not used everywhere

Reply:
maybe just a full `corev1.Probe` the timeouts can be different depending on the service. there might be also a difference needed between liveness and readiness ...